### ISSUE TITLE
Request: Make CustomModOption UNITS_NO_SUPPLY work again

### ISSUE THREAD

**Issue by UndeadDevel** [LoneGazebo/Community-Patch-DLL#5131]
_**1. Mod version** (i.e Date - 4/23):_
11-30-18

_**2. Mod list** (if using Vox Populi only, leave blank):_

VP (EUI)
City-States Leader for VP (v8)
List Enhancements for VP (v4.2)
Trade Opportunities for VP (v8.1)
Reforestation (v8) (slightly modified by me)
Civ-linked GG/GA (v11)
Enhanced Naval Warfare for VP (v1.6)
Enlightenment Era (VP) (v1.2)
Faster Aircraft Animations (v3)
Capture Great People (I neutered this one because of compatibility issues)
Espionage Mapping (v7) (modified by me)
Religious Settlers (v1)
Improved Air Recon (v1)
Missiles No Supply (v1)
More Unique Components for VP (v38)
Promotions Flags EUI (v1.1) (slightly modified by me)
Subs Ignore Borders (v3)
Subs Immune Under Ice (v2)
Destination (v4)
Filters in Trade Route Overview (VP) (v1.3) (slightly modified by me)
Improved City View (VP EUI) (v11.0)
Promotion Tree (v17)
Upgrade Tree (v9)
VP Promo Icons (v1.9)

_**3. Error description**:_
After the great success of reenabling the CustomModOption for CaptureEvents, I hope this one will work out as well.

The issue is that currently the CustomModOption UNITS_NO_SUPPLY does not seem to have an effect. I have tested this in my current game, which includes the mod [Missiles No Supply](http://www.picknmixmods.com/mods/CivV/Units/Missiles%20No%20Supply.html) (it sets the CustomModOption flag and sets the missile units as no-supply units) and unfortunately missiles do in fact use supply in my game. I have also tried setting another unit as no-supply but that didn't work either. This issue occurs both when I purchase and produce missiles.

Screenshots below show me buying a Guided Missile and my supply increasing (and the Supply popup before and after in case that's useful).

_**4. Steps to reproduce** (optional):_

1. Load VP game with mod [Missiles No Supply](http://www.picknmixmods.com/mods/CivV/Units/Missiles%20No%20Supply.html)
2. Produce or buy missile and watch supply rise

Supporting information:
Please note that you can attach .zip files by dragging-and-dropping them. If possible, zip up all supporting data and post that way.

1. Log files (always attach your Logs folder, located at My Documents/My Games/Sid Meier's Civilization 5. Make sure you have enabled logging before experiencing an error! Go here to find out how: http://forums.civfanatics.com/showthread.php?t=487482):

2. Save game (always attach a save that was made a turn before the error; located at My Documents/My Games/Sid Meier's Civilization 5/ModdedSaves):

3. CvMiniDump.dmp file (attach if experiencing a game crash. Located at Program Files/Steam/steamapps/common/Sid Meier's Civilization V):

4. Screenshots (optional):
![missiles1](https://user-images.githubusercontent.com/30495439/52028562-599d7180-2507-11e9-8c54-7ebc941996d6.jpg)
![missiles2](https://user-images.githubusercontent.com/30495439/52028570-60c47f80-2507-11e9-9e5d-37b7f4dddd53.jpg)
![missiles3](https://user-images.githubusercontent.com/30495439/52028576-6621ca00-2507-11e9-81d3-fc98fd9e42a3.jpg)
![missiles4](https://user-images.githubusercontent.com/30495439/52028586-6b7f1480-2507-11e9-9e01-4062bbaf19e1.jpg)

**Comment by LoneGazebo** [comment]
I see no reason for a conflict.

**Comment by UndeadDevel** [comment]
Sorry but it still doesn't work....just tested with newest version only (14-3-19) + Missiles No Supply mod and IGE to create the conditions. Giving missiles with IGE as well as buying them in the city does increase supply. I even modified the base settings, i.e. enabled UNITS_NO_SUPPLY in CP itself and included VP as a dependency in the MNS mod and still it increases supply when buying the Guided Missile.

**Comment by LoneGazebo** [comment]
Easiest solution - give Guided Missiles a promotion with <NoSupply>true</NoSupply>

That's the only way I see to do it, and it works in testing.

**Comment by UndeadDevel** [comment]
Hmm I can't seem to get it to work for some reason...the Database log tells me that there isn't a column "NoSupply" in the UnitPromotions table even if I turn on UNITS_NO_SUPPLY in CP directly; this is similar to the error when using the traditional Missiles No Supply mod (the log then notes that NoSupply doesn't exist as a column; I assume it means the units in that case).

It *should* work, since the NoSupply.sql file in the 'PNM Mods DB' folder inside CP adds those columns but I just can't get it to function and I don't know how to proceed further. If you have a little time, can you tell me if this works for you (it's the mod I made to replace Missiles No Supply, which now uses a promotion, as you suggested):
[Missiles No Supply for VP.zip](https://github.com/LoneGazebo/Community-Patch-DLL/files/2974622/Missiles.No.Supply.for.VP.zip)

And if not, how did you add the promotion successfully?

**Comment by supermerill** [comment]
maybe you should add cp as requirement for your mod, to make it load after cp?

**Comment by UndeadDevel** [comment]
> maybe you should add cp as requirement for your mod, to make it load after cp?
I did, you can check out the file I uploaded, it has both CP and CBO listed as dependencies. I also tried setting import on or off but it seems to me like the problem lies outside the mod mod anyway...I mean the error says that UnitPromotions doesn't have a "NoSupply" column, which is supposed to be added by VP when I turn on the relevant CustomModOption, which I tried doing in CP itself even, so that it's done before the mod mod even loads, but to no avail.

So yeah, I just don't know how to proceed from here and am super curious to know how @LoneGazebo got it to work.

**Comment by LoneGazebo** [comment]
Just add the relevant code to the CBO via copy paste.

**Comment by UndeadDevel** [comment]
>Just add the relevant code to the CBO via copy paste.

Thx, just tried that, but that results in validation failing for the PromotionChanges files in CBO. This is due to the line you suggested above:
`<NoSupply>true</NoSupply>`
because when I commented it out there was no issue anymore (other than nothing changing regarding No Supply not working).
Here is what I added in CBO:
`		<Row>
			<Type>PROMOTION_NO_SUPPLY</Type>
			<Description>TXT_KEY_PROMOTION_NO_SUPPLY</Description>
			<Help>TXT_KEY_PROMOTION_NO_SUPPLY_HELP</Help>
			<Sound>AS2D_IF_LEVELUP</Sound>
			<CannotBeChosen>true</CannotBeChosen>
                        <NoSupply>true</NoSupply>
			<PortraitIndex>44</PortraitIndex>
			<IconAtlas>extraPromo_Atlas</IconAtlas>
			<PediaType>PEDIA_ATTRIBUTES</PediaType>
			<PediaEntry>TXT_KEY_PROMOTION_NO_SUPPLY</PediaEntry>
		</Row>  `
in file PromotionChanges.xml in CBO and
`INSERT INTO UnitPromotions_UnitClasses
	(PromotionType, UnitClassType)
VALUES
	('PROMOTION_NO_SUPPLY', 'UNITCLASS_GUIDED_MISSILE'),
	('PROMOTION_NO_SUPPLY', 'UNITCLASS_NUCLEAR_MISSILE'),
	('PROMOTION_NO_SUPPLY', 'UNITCLASS_ROCKET_MISSILE');`
in PromotionChanges.sql in CBO.
I also changed the "Value" for the row with the "UNITS_NO_SUPPLY" Name attribute in the CustomModOptions.xml file of CP to "1", of course.

So I must not be understanding what you mean, but I don't see how I can do it any differently than I showed; can you clarify, please?

**Comment by LoneGazebo** [comment]
Try NoSupply as >1< not >true<

**Comment by LoneGazebo** [comment]
what does the invalidation explicitly state?

**Comment by UndeadDevel** [comment]
>Try NoSupply as >1< not >true<

still not working.
Relevant part from Database.log is [here](https://pastebin.com/y8SKyvm8)

**Comment by LoneGazebo** [comment]
Is the INSERT INTO section in XML? Because that's SQL, it has to be in the promotionchanges.sql file.

Copy these lines: 

ALTER TABLE Units
  ADD NoSupply INTEGER DEFAULT 0;
  
ALTER TABLE UnitPromotions
  ADD NoSupply INTEGER DEFAULT 0;

into the \MODS\(1) Community Patch (v 89)\Core Files\Core Tables\CoreTableEntries.sql file (Anywhere is fine).

**Comment by UndeadDevel** [comment]
>Is the INSERT INTO section in XML? Because that's SQL, it has to be in the promotionchanges.sql file.

No that's in the sql file, the other stuff is in the xml file.

Tried inserting the code you posted and that makes the validation failure disappear but the functionality is still missing...trying again with "true" instead of "1".

**Comment by UndeadDevel** [comment]
Nope, still not working even after I changed the TXT_KEYs to remove all error lines in the log that come from the changes and pass validation.

**Comment by LoneGazebo** [comment]
WTH. I’ll look.
